src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,288 @@ namespace Microsoft.CodeAnalysis
     /// </summary>
     public class ProjectDependencyGraph
     {
-        private readonly ImmutableArray<ProjectId> _projectIds;
+        private readonly ImmutableHashSet<ProjectId> _projectIds;
         private readonly ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> _referencesMap;
 
         // guards lazy computed data
         private readonly NonReentrantLock _dataLock = new NonReentrantLock();
 
-        // these are computed fully on demand
+        // These are computed fully on demand. null or ImmutableArray.IsDefault indicates the item needs to be realized
         private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> _lazyReverseReferencesMap;
         private ImmutableArray<ProjectId> _lazyTopologicallySortedProjects;
         private ImmutableArray<IEnumerable<ProjectId>> _lazyDependencySets;
 
-        // these accumulate results on demand
-        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> _transitiveReferencesMap = ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>.Empty;
-        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> _reverseTransitiveReferencesMap = ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>.Empty;
+        // These accumulate results on demand. They are never null, but a missing key/value pair indicates it needs to be computed.
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> _transitiveReferencesMap;
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> _reverseTransitiveReferencesMap;
```

Comment:
Do we need all these dictionaries? Can't we instead use some representation like `Node { ProjectId Id; ImmutableArray<Node> References, ImmutableArray<Node> ReferencedBy; }` with these arrays sorted, then use binary search for lookup?

Reply:
Or I guess one ImmutableDictionary<ProjectId, (ProjectId[] References, ProjectId[] ReferencedBy)> might have lower incremental update overhead.

---
In reply to: [210750520](https://github.com/dotnet/roslyn/pull/28441#discussion_r210750520) [](ancestors = 210750520)

Reply:
Filed [https://github.com/dotnet/roslyn/issues/29430](https://github.com/dotnet/roslyn/issues/29430) to follow up

---
In reply to: [210753567](https://github.com/dotnet/roslyn/pull/28441#discussion_r210753567) [](ancestors = 210753567,210750520)